File: packs/telemetry/behavior_pack/scripts/domain/metrics.js
/**
 * Opt-in metric policy (docs/telemetry-metrics.md).
 *
 * Pure decisions only: which metrics exist, what a command means, and which
 * map keys may be stored. Persistence lives in adapters/store.js and the
 * counting lives in main.js, so this module can be reasoned about — and
 * tested — without a runtime.
 */

/** Every metric this pack knows. A name outside this list is refused. */
export const METRICS = ["itemUse"];

/**
 * A Minecraft namespaced identifier: lowercase namespace, colon, path.
 *
 * The shape is the privacy boundary, not a formality. A player-authored name
 * carries capitals, spaces or formatting codes and cannot match, so a custom
 * item name can never become a map key.
 */
const IDENTIFIER = /^[a-z0-9_]+:[a-z0-9_./-]+$/;
const IDENTIFIER_MAX_LENGTH = 112;

export function emptyMetrics() {
  return Object.fromEntries(METRICS.map((name) => [name, false]));
}

/**
 * Read persisted metric state, defaulting anything unrecognized to disabled.
 *
 * A corrupt or partial value must not enable collection by accident, so every
 * unknown name is dropped and every non-boolean reads as `false`.
 */
export function parseMetrics(value) {
  const source = value !== null && typeof value === "object" && !Array.isArray(value) ? value : {};
  return Object.fromEntries(METRICS.map((name) => [name, source[name] === true]));
}

/**
 * Interpret one `bedrock_telemetry:metrics` command.
 *
 * Accepts `enable <metric>`, `disable <metric>` and `status`. Returns the
 * resulting state plus whether it differs from the current one; an
 * unrecognized command returns an error and changes nothing.
 */
export function applyMetricCommand(current, message) {
  const [verb = "", name = ""] = String(message ?? "").trim().split(/\s+/);
  const metrics = parseMetrics(current);
  if (verb === "status" && !name) return { metrics, changed: false, error: null };
  if (verb !== "enable" && verb !== "disable") return { metrics, changed: false, error: `unknown metric command: ${verb || "(empty)"}` };
  if (!METRICS.includes(name)) return { metrics, changed: false, error: `unknown metric: ${name || "(none)"}` };
  const enabled = verb === "enable";
  return { metrics: { ...metrics, [name]: enabled }, changed: metrics[name] !== enabled, error: null };
}

/** Return the identifier when it may be stored as a map key, otherwise null. */
export function metricKey(value) {
  if (typeof value !== "string" || value.length > IDENTIFIER_MAX_LENGTH) return null;
  return IDENTIFIER.test(value) ? value : null;
}

File: apps/server/controlplane/src/telemetry/metrics.py
"""Opt-in telemetry metrics — the manager's half of the contract.

``docs/telemetry-metrics.md`` decides what the pack may collect; this module
owns the manager-side allowlist of metric names and the shape of the state the
pack reports back. Nothing here talks to the console or the database.
"""

from __future__ import annotations

import json
import re
from typing import Any

#: Every metric the manager may switch. A name outside this set is refused
#: before anything reaches the Bedrock console.
METRICS = ("itemUse", "blockInteractions", "entityInteractions", "containerInteractions")


#: A Minecraft namespaced identifier, the only shape an opt-in metric map may
#: use as a key. Player-authored text carries capitals, spaces or formatting
#: codes and cannot match, which is what keeps a custom name out of the data.
NAMESPACED_IDENTIFIER = re.compile(r"^[a-z0-9_]+:[a-z0-9_./-]+$")
IDENTIFIER_MAX_LENGTH = 112


def is_metric_key(value: Any) -> bool:
    """Return whether a value may be stored as a metric map key."""
    return (
        isinstance(value, str)
        and len(value) <= IDENTIFIER_MAX_LENGTH
        and bool(NAMESPACED_IDENTIFIER.match(value))
    )


class UnknownMetric(ValueError):
    """Raised for a metric name outside the allowlist."""

    def __init__(self, metric: Any) -> None:
        super().__init__(f"Métrica de telemetria desconhecida: {metric}")
        self.metric = metric


def validate(metric: Any) -> str:
    """Return the metric name, or refuse it."""
    if not isinstance(metric, str) or metric not in METRICS:
        raise UnknownMetric(metric)
    return metric


def normalize(value: Any) -> dict[str, bool]:
    """Read a reported metric map, defaulting anything unrecognized to off.

    A metric the pack does not report is *not enabled*: an older pack, a
    truncated payload or a corrupt property must never read as opted in.
    """
    source = value if isinstance(value, dict) else {}
    return {metric: source.get(metric) is True for metric in METRICS}


def loads(raw: Any) -> dict[str, bool]:
    """Read the metric map persisted in the telemetry state table."""
    try:
        return normalize(json.loads(raw) if isinstance(raw, str) and raw else None)
    except ValueError:
        return normalize(None)


def dumps(value: Any) -> str:
    return json.dumps(normalize(value), sort_keys=True)
